### PR TITLE
Introduce customizable Decorum.call()

### PR DESCRIPTION
**Pull request by benoitbryon**
Refs #23. The `call()` method makes it easy to pre-process input arguments or post-process results.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -18,7 +18,7 @@ Typical usage looks like this:
    >>> class my_decorator(Decorum):
    ...    def wraps(self, f):
    ...        print("I'm returning the function! You can keep it!")
-   ...        return f
+   ...        return super(my_decorator, self).wraps(f)
```
Comment (benoitbryon) [inline README.rst]: The way I refactored the code, we have to call super().wraps() in order to take advantage of Decorum's base features: assign `self._wrapped` and assign `__name__` and `__doc__`.

File: README.rst [hunk 2]
```diff
@@ -66,30 +65,33 @@ Here is a slightly fancier example:
...
    ...         else:
-   ...             def newf():
-   ...                 print('wut')
-   ...         return newf
+   ...             newf = lambda: 'wut'
+   ...         return super(fancy, self).wraps(newf)
+   ...
+   ...     def call(self, *args, **kwargs):
+   ...         result = super(fancy, self).call(*args, **kwargs)
+   ...         return result.upper()
```
Comment (benoitbryon) [inline README.rst]: Here is an example of result post-processing.

File: README.rst [hunk 3]
```diff
@@ -66,30 +65,33 @@ Here is a slightly fancier example:
...
    >>> class fancy(Decorum):
    ...     def init(self, arg=None):
+   ...         super(fancy, self).init()
    ...         self.arg = arg
    ...
    ...     def wraps(self, f):
    ...         if self.arg:
-   ...             def newf():
-   ...                 print(self.arg)
+   ...             newf = lambda: self.arg
```
Comment (benoitbryon) [inline README.rst]: I shortened the code with a lambda function... This change is not related to the feature request (a.k.a. "call" method).

Comment (benoitbryon) [inline README.rst]: I also uses `return` instead of `print`, so that I can easily intercept and post-process the result.

File: README.rst [hunk 4]
```diff
@@ -66,30 +65,33 @@ Here is a slightly fancier example:
 
    >>> class fancy(Decorum):
    ...     def init(self, arg=None):
+   ...         super(fancy, self).init()
```
Comment (benoitbryon) [inline README.rst]: This use of super() is not mandatory here since parent class does nothing in init(). But I think it is a good practice.

File: README.rst [hunk 5]
```diff
@@ -66,30 +65,33 @@ Here is a slightly fancier example:
...
+   ...             newf = lambda: self.arg
    ...         else:
-   ...             def newf():
-   ...                 print('wut')
-   ...         return newf
+   ...             newf = lambda: 'wut'
+   ...         return super(fancy, self).wraps(newf)
+   ...
+   ...     def call(self, *args, **kwargs):
+   ...         result = super(fancy, self).call(*args, **kwargs)
```
Comment (benoitbryon) [inline README.rst]: Here, super() just runs the wrapped function and returns the result.

File: decorum/decorum.py [hunk 6]
```diff
@@ -26,40 +29,40 @@ def __init__(self, *args, **kwargs):
         #: directly to the matching attributes on the decorator.
         self.assigned = functools.WRAPPER_ASSIGNMENTS
         if 'assigned' in kwargs:
-            self.assigned = kwargs['assigned']
+            self.assigned = kwargs.pop('assigned')
```
Comment (benoitbryon) [inline decorum/decorum.py]: I used "pop()" so that init() doesn't receive the "assigned" argument.

File: decorum/decorum.py [hunk 7]
```diff
@@ -26,40 +29,40 @@ def __init__(self, *args, **kwargs):
...
             # used as decorator without being called
             self.init()
-            self._wrapped = self.__call__(args[0])
+            self.wraps(args[0])
         else:
             # used as decorator and called
             self.init(*args, **kwargs)
 
-    def __call__(self, f=None, *args, **kwargs):
+    def __call__(self, *args, **kwargs):
```
Comment (benoitbryon) [inline decorum/decorum.py]: I replaced `f=None` with the `args[0]` like in `__init__()`.

File: decorum/decorum.py [hunk 8]
```diff
@@ -26,40 +29,40 @@ def __init__(self, *args, **kwargs):
...
         Uses :meth:`wrap` to handle decoration. Restores :attr:`__doc__` and
         :attr:`__name__`.
 
         """
-        if not callable(f):
-            if f:
-                return self._wrapped(f, *args, **kwargs)
-            else:
-                return self._wrapped(*args, **kwargs)
+        if self._wrapped:
```
Comment (benoitbryon) [inline decorum/decorum.py]: This is a partial implementation of #24.

File: decorum/decorum.py [hunk 9]
```diff
@@ -26,40 +29,40 @@ def __init__(self, *args, **kwargs):
...
+            return self.call(*args, **kwargs)
         else:
-            wrapped = self.wraps(f)
-            return wrapped
+            f = args[0]
+            return self.wraps(f)
 
     def wraps(self, f):
         """Wraps the function and returns it"""
+        self._wrapped = f
```
Comment (benoitbryon) [inline decorum/decorum.py]: I wonder about moving this line in both `__init__` and `__call__`, so that, if users forget to call `super(subclass, self).wraps(function)`, they lose the update_wrapper feature, but still have the `self._wrapped` attribute.

File: decorum/decorum.py [hunk 10]
```diff
@@ -26,40 +29,40 @@ def __init__(self, *args, **kwargs):
...
         functools.update_wrapper(self, f, self.assigned or (), ())
         return self
 
-    def init(self, *args, **kwargs):
+    def init(self):
```
Comment (benoitbryon) [inline decorum/decorum.py]: I think `init()` signature should tell exactly the decorator's own options.
I mean:
* if the decorator doesn't accept configuration, then `def init(self):`.
* if the decorator requires one positional argument, then `def init(self, this_is_required):`.
* if the decorator accepts one argument with a default value, then `def init(self, this_is_optional=True):`.
* if the decorator accepts only keyword arguments, then `def init(self, **kwargs):`
* ...

**Comment by benoitbryon** [comment]
I commented the diff in order to point out and explain most important changes.
I think I'm starting to introduce some backward incompatibilities here :( Mostly around the use of super().

**Comment by zeekay** [comment]
Looks good!
